@xen-orchestra/xapi/vm.mjs

```diff
@@ -599,6 +599,32 @@ class Vm {
     }
   }
 
+  async leafCoalesce($defer, vmRef) {
+    const vmUuid = await this.getField('VM', vmRef, 'uuid')
+    try {
+      await this.callAsync('VM.suspend', vmRef)
+      $defer(() => this.resumeVm(vmUuid))
+    } catch (error) {
+      if (error.code !== 'VM_BAD_POWER_STATE') {
+        throw error
+      }
+
+      const powerState = error.params[2].toLowerCase()
+      if (powerState !== 'halted' && powerState !== 'suspended') {
+        throw error
+      }
+    }
+
+    // result can be: `Success` or `VM has no leaf-coalesceable VDIs`
+    const result = await this.callAsync('host.call_plugin', this.pool.master, 'coalesce-leaf', 'leaf-coalesce', {
+      vm_uuid: vmUuid,
+    })
+
+    if (result === 'VM has no leaf-coalesceable VDIs') {
```

Comment:
Testing success is probably more robust (e.g. if other errors are implemented in the future.